SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialLatencyTracer.h

```diff
@@ -0,0 +1,116 @@
+	//		- continue the trace using `ContinueLatencyTrace`, again tagging it against another Actor's RPC
+	//		- or end the trace using `EndLatencyTrace`
+	//
+	//////////////////////////////////////////////////////////////////////////
+
+	USpatialLatencyTracer();
+
+	// Front-end exposed, allows users to register, start, continue, and end traces
+	UFUNCTION(BlueprintCallable, Category = "SpatialOS", meta = (WorldContext = "WorldContextObject"))
+	static void RegisterProject(UObject* WorldContextObject, const FString& ProjectId);
+
+	UFUNCTION(BlueprintCallable, Category = "SpatialOS", meta = (WorldContext = "WorldContextObject"))
+	static bool BeginLatencyTrace(UObject* WorldContextObject, const AActor* Actor, const FString& FunctionName, const FString& TraceDesc);
+
+	UFUNCTION(BlueprintCallable, Category = "SpatialOS", meta = (WorldContext = "WorldContextObject"))
+	static bool ContinueLatencyTrace(UObject* WorldContextObject, const AActor* Actor, const FString& FunctionName, const FString& TraceDesc);
+
+	UFUNCTION(BlueprintCallable, Category = "SpatialOS", meta = (WorldContext = "WorldContextObject"))
+	static bool EndLatencyTrace(UObject* WorldContextObject);
+
+	static const TraceKey ActiveTraceKey = 0;
+	static const TraceKey InvalidTraceKey = -1;
+
+#if TRACE_LIB_ACTIVE
+
+	// Internal GDK usage, shouldn't be used by game code
+	static USpatialLatencyTracer* GetTracer(UObject* WorldContextObject);
+
+	bool IsValidKey(TraceKey Key);
+	TraceKey GetTraceKey(const UObject* Obj, const UFunction* Function);
+
+	void WriteToLatencyTrace(const TraceKey Key, const FString& TraceDesc);
+	void EndLatencyTrace(const TraceKey Key, const FString& TraceDesc);
+
+	void WriteTraceToSchemaObject(const TraceKey Key, Schema_Object* Obj, const Schema_FieldId FieldId);
+	TraceKey ReadTraceFromSchemaObject(Schema_Object* Obj, const Schema_FieldId FieldId);
+
+	void SetWorkerId(const FString& NewWorkerId) { WorkerId = NewWorkerId; }
+	void ResetWorkerId() { WorkerId = TEXT("Undefined"); }
+
+private:
+
+	using ActorFuncKey = TPair<const AActor*, const UFunction*>;
+	using TraceSpan = improbable::trace::Span;
+
+	bool BeginLatencyTrace_Internal(const AActor* Actor, const FString& FunctionName, const FString& TraceDesc);
+	bool ContinueLatencyTrace_Internal(const AActor* Actor, const FString& FunctionName, const FString& TraceDesc);
+	bool EndLatencyTrace_Internal();
+
+	TraceKey CreateNewTraceEntry(const AActor* Actor, const FString& FunctionName);
+	TraceSpan* GetActiveTrace();
+
+	void WriteKeyFrameToTrace(const TraceSpan* Trace, const FString& TraceDesc);
+	FString FormatMessage(const FString& Message) const;
+
+
+	FString WorkerId;
+
+	FCriticalSection Mutex; // This mutex is to protect modifications to the containers below
+	TMap<ActorFuncKey, TraceKey> TrackingTraces;
+	TMap<TraceKey, TraceSpan> TraceMap;
+
+public:
+
+#endif // TRACE_LIB_ACTIVE
+
+	UFUNCTION(BlueprintCallable, Category = "SpatialOS")
+	static void SendTestTrace();
```

Comment:
What is this for?

Reply:
For sending a fully contained test trace. Useful for ensuring the trace library is working on a single platform/worker. Will add an explanation comment. Are you okay with it staying in?

Reply:
Sorry, I meant to be run as a unit test or as a thing that needs to happen at the beginning of each deployment.

Reply:
It's for debugging